### PR TITLE
[ios][battery] Migrate module to Sweet API

### PR DESCRIPTION
**Pull request by fobos531**
# Why

Continuation of migrating expo modules to the Sweet API

# How

- routine work to migrate the module to Sweet API
- pretty much 1:1 migration from the old ObjC module to the Sweet API, using other modules (namely `expo-clipboard`) as inspiration

# Test Plan

Tested NCL example on:
- iPhone 8 simulator - iOS 16.0
- iPhone 12 physical device - iOS 16.0.2

From my testing everything seems to be working fine. The only thing I'm unsure of is the `batteryState` method since iOS was always returning `unknown` battery state for my physical device. 

This was a ton of fun to do! I'm not very deeply experienced with Swift so my code might be off in terms of the module lifecycle parts (`onCreate`, `onDestroy`, `onStartObserving`, `onStopObserving`) - open to feedback on that front.

# Checklist

- [X] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- [X] Conforms with the [Documentation Writing Style Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)
- [X] This diff will work correctly for `expo prebuild` & EAS Build (eg: updated a module plugin).

### CODE REVIEW

File: packages/expo-battery/src/ExpoBattery.ts [hunk 1]
```diff
@@ -1,2 +1,2 @@
-import { NativeModulesProxy } from 'expo-modules-core';
-export default NativeModulesProxy.ExpoBattery;
+import { requireNativeModule } from 'expo-modules-core';
+export default requireNativeModule('ExpoBattery');
```
Comment (fobos531) [inline packages/expo-battery/src/ExpoBattery.ts:2]: may require https://github.com/expo/expo/pull/19482 to fully work correctly

Comment (tsapeta) [reply packages/expo-battery/src/ExpoBattery.ts:2]: Hah, yeah, your PR reminded me to do that 😂 Thanks!

**Comment by fobos531** [comment]
Update on the `batteryState` issue - I managed to fix this, it is now being reported correctly. However, I'm not completely sure about the listeners - to my knowledge they worked perfectly fine last time I tested them, but now they do not send events for whatever reason, might be something related to my local setup 😅

**Comment by tsapeta** [comment]
@fobos531 I've merged #19482 to main, could you try to rebase your branch and test events again?

**Comment by fobos531** [comment]
> @fobos531 I've merged #19482 to main, could you try to rebase your branch and test events again?

sure, i'll test as soon as I'm able to and report back

File: packages/expo-battery/ios/BatteryModule.swift [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+    return true
+  #endif
+}
+
+enum BatteryState: Int {
+   case unknown = 0
+   case unplugged = 1
+   case charging = 2
+   case full = 3
+ }
```
Comment (tsapeta) [inline packages/expo-battery/ios/BatteryModule.swift:113]:
```suggestion
}
```

File: packages/expo-battery/ios/BatteryModule.swift [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+    AsyncFunction("getBatteryLevelAsync") { () -> Float in
+      return UIDevice.current.batteryLevel
+    }
+
+    AsyncFunction("isLowPowerModeEnabledAsync") { () -> Bool in
+      return ProcessInfo.processInfo.isLowPowerModeEnabled
+    }
+
+    AsyncFunction("getBatteryStateAsync") { () -> Int in
+      switch UIDevice.current.batteryState {
```
Comment (tsapeta) [inline packages/expo-battery/ios/BatteryModule.swift:26]: Looks like `UIDevice.BatteryState` has the same cases and values as ours `BatteryState`, so maybe we can just return `UIDevice.current.batteryState.rawValue` here? 🤔 If so, then please also add a comment that this type is in-sync with our TS enum.

Comment (fobos531) [reply packages/expo-battery/ios/BatteryModule.swift:26]: If that's the case, then yeah - we can definitely do that. I agree on the comment remark, we should definitely include so we're aware of what we do if the enums ever become "out-of-sync"

**Review by tsapeta — APPROVED** [review]
Great work, as always! 😉 Thank you! 🙏

**Pull request merged**